src/lib.rs

```diff
@@ -1833,17 +1833,43 @@ impl<K: KVStore + Sync + Send + 'static> Node<K> {
 
 	/// Retrieves a list of known peers.
 	pub fn list_peers(&self) -> Vec<PeerDetails> {
-		let active_connected_peers: Vec<PublicKey> =
-			self.peer_manager.get_peer_node_ids().iter().map(|p| p.0).collect();
-		self.peer_store
-			.list_peers()
-			.iter()
-			.map(|p| PeerDetails {
+		let mut peers = Vec::new();
+
+		// First add all connected peers, preferring to list the connected address if available.
+		let connected_peers = self.peer_manager.get_peer_node_ids();
+		for (node_id, con_addr_opt) in connected_peers {
+			let stored_peer = self.peer_store.get_peer(&node_id);
+			let stored_addr_opt = stored_peer.as_ref().map(|p| p.address.clone());
+			let address = match (con_addr_opt, stored_addr_opt) {
+				(Some(con_addr), Some(_stored_addr)) => NetAddress(con_addr),
+				(None, Some(stored_addr)) => stored_addr,
+				(Some(con_addr), None) => NetAddress(con_addr),
+				(None, None) => continue,
+			};
+
+			let is_persisted = stored_peer.is_some();
+			let is_connected = true;
+			let details = PeerDetails { node_id, address, is_persisted, is_connected };
+			peers.push(details);
+		}
+
+		// Now add all known-but-offline peers, too.
+		for p in self.peer_store.list_peers() {
+			if peers.iter().find(|d| d.node_id == p.node_id).is_some() {
+				continue;
+			}
```

Comment:
Seems we can be more efficient here by only iterating through the initial set of connected peers instead of the growing list.